File: sdks/java/core/src/test/java/com/google/cloud/dataflow/sdk/transforms/MaxTest.java
/*
 * Copyright (C) 2015 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */

package com.google.cloud.dataflow.sdk.transforms;

import static com.google.cloud.dataflow.sdk.TestUtils.checkCombineFn;
import static org.junit.Assert.assertEquals;

import com.google.common.collect.Lists;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.JUnit4;

/**
 * Tests for Max.
 */
@RunWith(JUnit4.class)
public class MaxTest {
  @Test
  public void testMeanGetNames() {
    assertEquals("Max.Globally", Max.integersGlobally().getName());
    assertEquals("Max.Globally", Max.doublesGlobally().getName());
    assertEquals("Max.Globally", Max.longsGlobally().getName());
    assertEquals("Max.PerKey", Max.integersPerKey().getName());
    assertEquals("Max.PerKey", Max.doublesPerKey().getName());
    assertEquals("Max.PerKey", Max.longsPerKey().getName());
  }

  @Test
  public void testMaxIntegerFn() {
    checkCombineFn(
        new Max.MaxIntegerFn(),
        Lists.newArrayList(1, 2, 3, 4),
        4);
  }

  @Test
  public void testMaxLongFn() {
    checkCombineFn(
        new Max.MaxLongFn(),
        Lists.newArrayList(1L, 2L, 3L, 4L),
        4L);
  }

  @Test
  public void testMaxDoubleFn() {
    checkCombineFn(
        new Max.MaxDoubleFn(),
        Lists.newArrayList(1.0, 2.0, 3.0, 4.0),
        4.0);
  }
}

File: sdks/go/pkg/beam/transforms/stats/max_test.go
// Licensed to the Apache Software Foundation (ASF) under one or more
// contributor license agreements.  See the NOTICE file distributed with
// this work for additional information regarding copyright ownership.
// The ASF licenses this file to You under the Apache License, Version 2.0
// (the "License"); you may not use this file except in compliance with
// the License.  You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package stats

import (
	"testing"

	"github.com/apache/beam/sdks/v2/go/pkg/beam"
	"github.com/apache/beam/sdks/v2/go/pkg/beam/register"
	"github.com/apache/beam/sdks/v2/go/pkg/beam/testing/passert"
	"github.com/apache/beam/sdks/v2/go/pkg/beam/testing/ptest"
)

func init() {
	register.Function2x1(kvToStudent)
	register.Function1x2(studentToKV)
}

type student struct {
	Name  string
	Grade float64
}

func kvToStudent(n string, g float64) student {
	return student{n, g}
}

func studentToKV(s student) (string, float64) {
	return s.Name, s.Grade
}

// TestMaxInt verifies that Max works correctly for ints.
func TestMaxInt(t *testing.T) {
	tests := []struct {
		in  []int
		exp []int
	}{
		{
			[]int{1, -4, 3},
			[]int{3},
		},
		{
			[]int{7, 11, 7, 5, 10},
			[]int{11},
		},
		{
			[]int{0, -2, -10},
			[]int{0},
		},
		{
			[]int{-9},
			[]int{-9},
		},
	}

	for _, test := range tests {
		p, s, in, exp := ptest.CreateList2(test.in, test.exp)
		passert.Equals(s, Max(s, in), exp)

		if err := ptest.Run(p); err != nil {
			t.Errorf("Max(%v) != %v: %v", test.in, test.exp, err)
		}
	}
}

// TestMaxFloat verifies that Max works correctly for float64s.
func TestMaxFloat(t *testing.T) {
	tests := []struct {
		in  []float64
		exp []float64
	}{
		{
			[]float64{1, -2, 3.5},
			[]float64{3.5},
		},
		{
			[]float64{0, -99.99, 1, 1},
			[]float64{1},
		},
		{
			[]float64{5.67890},
			[]float64{5.6789},
		},
	}

	for _, test := range tests {
		p, s, in, exp := ptest.CreateList2(test.in, test.exp)
		passert.Equals(s, Max(s, in), exp)

		if err := ptest.Run(p); err != nil {
			t.Errorf("Max(%v) != %v: %v", test.in, test.exp, err)
		}
	}
}

// TestMaxKeyed verifies that Max works correctly for KV values.
func TestMaxKeyed(t *testing.T) {
	tests := []struct {
		in  []student
		exp []student
	}{
		{
			[]student{{"alpha", 1}, {"beta", 4}, {"charlie", 3.5}},
			[]student{{"alpha", 1}, {"beta", 4}, {"charlie", 3.5}},
		},
		{
			[]student{{"alpha", 1}},
			[]student{{"alpha", 1}},
		},
		{
			[]student{{"alpha", 1}, {"alpha", -4}, {"beta", 4}, {"charlie", 0}, {"charlie", 5.5}},
			[]student{{"alpha", 1}, {"beta", 4}, {"charlie", 5.5}},
		},
	}

	for _, test := range tests {
		p, s, in, exp := ptest.CreateList2(test.in, test.exp)
		kv := beam.ParDo(s, studentToKV, in)
		max := MaxPerKey(s, kv)
		maxStudent := beam.ParDo(s, kvToStudent, max)
		passert.Equals(s, maxStudent, exp)

		if err := ptest.Run(p); err != nil {
			t.Errorf("Max(%v) != %v: %v", test.in, test.exp, err)
		}
	}
}
